File: ios/MenuMakerUITests/PageObjects/SellerOrdersPage.swift
//
//  SellerOrdersPage.swift
//  MenuMakerUITests
//
//  Page Object for Seller Orders Management Screen
//

import XCTest

struct SellerOrdersPage {
    let app: XCUIApplication

    // MARK: - Elements

    var orderCards: XCUIElementQuery {
        app.scrollViews.otherElements.matching(identifier: "OrderCard")
    }

    var firstOrder: XCUIElement {
        orderCards.firstMatch
    }

    var newOrdersTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'new' OR label CONTAINS[c] 'pending'")).firstMatch
    }

    var activeOrdersTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'active' OR label CONTAINS[c] 'preparing'")).firstMatch
    }

    var completedOrdersTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'completed' OR label CONTAINS[c] 'history'")).firstMatch
    }

    // Order detail elements
    var acceptButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'accept'")).firstMatch
    }

    var rejectButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'reject'")).firstMatch
    }

    var markPreparingButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'preparing' OR label CONTAINS[c] 'start'")).firstMatch
    }

    var markReadyButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'ready' OR label CONTAINS[c] 'complete'")).firstMatch
    }

    var orderIdLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label MATCHES '#\\\\d+'")).firstMatch
    }

    var customerNameLabel: XCUIElement {
        app.staticTexts.matching(identifier: "CustomerName").firstMatch
    }

    var orderTotalLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹'")).firstMatch
    }

    var emptyStateMessage: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'no orders' OR label CONTAINS[c] 'no new orders'")).firstMatch
    }

    var rejectionReasonField: XCUIElement {
        app.textViews.matching(NSPredicate(format: "placeholderValue CONTAINS[c] 'reason'")).firstMatch
    }

    var confirmRejectButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label == 'Reject' OR label == 'Confirm'")).firstMatch
    }

    // MARK: - Actions

    @discardableResult
    func tapFirstOrder() -> SellerOrdersPage {
        firstOrder.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func switchToNewOrders() -> SellerOrdersPage {
        if newOrdersTab.exists {
            newOrdersTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToActiveOrders() -> SellerOrdersPage {
        if activeOrdersTab.exists {
            activeOrdersTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToCompletedOrders() -> SellerOrdersPage {
        if completedOrdersTab.exists {
            completedOrdersTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func acceptOrder() -> SellerOrdersPage {
        acceptButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func rejectOrder(reason: String? = nil) -> SellerOrdersPage {
        rejectButton.tap()

        if let reason = reason, rejectionReasonField.waitForExistence(timeout: 1) {
            rejectionReasonField.tap()
            rejectionReasonField.typeText(reason)
        }

        if confirmRejectButton.waitForExistence(timeout: 1) {
            confirmRejectButton.tap()
        }

        sleep(1)
        return self
    }

    @discardableResult
    func markAsPreparing() -> SellerOrdersPage {
        markPreparingButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func markAsReady() -> SellerOrdersPage {
        markReadyButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func pullToRefresh() -> SellerOrdersPage {
        let firstCell = app.scrollViews.firstMatch
        let start = firstCell.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.2))
        let end = firstCell.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.8))
        start.press(forDuration: 0, thenDragTo: end)
        sleep(2)
        return self
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> SellerOrdersPage {
        XCTAssertTrue(orderCards.firstMatch.waitForExistence(timeout: timeout) ||
                     emptyStateMessage.waitForExistence(timeout: timeout),
                     "Orders screen should be displayed")
        return self
    }

    @discardableResult
    func assertOrdersDisplayed() -> SellerOrdersPage {
        XCTAssertTrue(firstOrder.exists, "Orders should be displayed")
        return self
    }

    @discardableResult
    func assertEmptyState() -> SellerOrdersPage {
        XCTAssertTrue(emptyStateMessage.exists, "Empty state should be displayed")
        return self
    }

    @discardableResult
    func assertOrderCount(_ expectedCount: Int) -> SellerOrdersPage {
        let actualCount = orderCards.count
        XCTAssertEqual(actualCount, expectedCount, "Should have \(expectedCount) orders, found \(actualCount)")
        return self
    }

    @discardableResult
    func assertOrderDetailDisplayed(timeout: TimeInterval = 2) -> SellerOrdersPage {
        XCTAssertTrue(orderIdLabel.waitForExistence(timeout: timeout) ||
                     orderTotalLabel.waitForExistence(timeout: timeout),
                     "Order detail should be displayed")
        return self
    }

    @discardableResult
    func assertAcceptButtonVisible() -> SellerOrdersPage {
        XCTAssertTrue(acceptButton.exists, "Accept button should be visible")
        return self
    }

    @discardableResult
    func assertMarkPreparingButtonVisible() -> SellerOrdersPage {
        XCTAssertTrue(markPreparingButton.exists, "Mark preparing button should be visible")
        return self
    }

    @discardableResult
    func assertMarkReadyButtonVisible() -> SellerOrdersPage {
        XCTAssertTrue(markReadyButton.exists, "Mark ready button should be visible")
        return self
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/SellerOrdersPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Seller Orders Management Screen
 * Provides fluent API for seller order management interactions
 */
class SellerOrdersPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val orderCards = composeTestRule.onAllNodesWithTag("OrderCard")
    private val newOrdersTab = composeTestRule.onNode(
        hasText("new", substring = true, ignoreCase = true) or
        hasText("pending", substring = true, ignoreCase = true)
    )
    private val activeOrdersTab = composeTestRule.onNode(
        hasText("active", substring = true, ignoreCase = true) or
        hasText("preparing", substring = true, ignoreCase = true)
    )
    private val completedOrdersTab = composeTestRule.onNode(
        hasText("completed", substring = true, ignoreCase = true) or
        hasText("history", substring = true, ignoreCase = true)
    )

    // Order action buttons
    private val acceptButton = composeTestRule.onNodeWithText("accept", substring = true, ignoreCase = true)
    private val rejectButton = composeTestRule.onNodeWithText("reject", substring = true, ignoreCase = true)
    private val markPreparingButton = composeTestRule.onNode(
        hasText("preparing", substring = true, ignoreCase = true) or
        hasText("start", substring = true, ignoreCase = true)
    )
    private val markReadyButton = composeTestRule.onNode(
        hasText("ready", substring = true, ignoreCase = true) or
        hasText("complete", substring = true, ignoreCase = true)
    )

    // Order detail elements
    private val orderIdLabel = composeTestRule.onNodeWithTag("order-id")
    private val customerNameLabel = composeTestRule.onNodeWithTag("CustomerName")
    private val orderTotalLabel = composeTestRule.onNode(hasText("₹", substring = true))
    private val emptyStateMessage = composeTestRule.onNode(
        hasText("no orders", substring = true, ignoreCase = true) or
        hasText("no new orders", substring = true, ignoreCase = true)
    )
    private val rejectionReasonField = composeTestRule.onNodeWithTag("rejection-reason-field")
    private val confirmRejectButton = composeTestRule.onNodeWithText("Reject") or composeTestRule.onNodeWithText("Confirm")

    // Actions
    fun tapFirstOrder(): SellerOrdersPage {
        if (orderCards.fetchSemanticsNodes().isNotEmpty()) {
            orderCards[0].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun switchToNewOrders(): SellerOrdersPage {
        newOrdersTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToActiveOrders(): SellerOrdersPage {
        activeOrdersTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToCompletedOrders(): SellerOrdersPage {
        completedOrdersTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun acceptOrder(): SellerOrdersPage {
        acceptButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun rejectOrder(reason: String? = null): SellerOrdersPage {
        rejectButton.performClick()

        if (reason != null && rejectionReasonField.fetchSemanticsNode(false) != null) {
            rejectionReasonField.performTextInput(reason)
        }

        val confirmBtn = composeTestRule.onNode(
            hasText("reject", ignoreCase = true) or hasText("confirm", ignoreCase = true)
        )
        if (confirmBtn.fetchSemanticsNode(false) != null) {
            confirmBtn.performClick()
        }

        Thread.sleep(1000)
        return this
    }

    fun markAsPreparing(): SellerOrdersPage {
        markPreparingButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun markAsReady(): SellerOrdersPage {
        markReadyButton.performClick()
        Thread.sleep(1000)
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): SellerOrdersPage {
        composeTestRule.waitUntil(timeoutMillis = 2000) {
            orderCards.fetchSemanticsNodes().isNotEmpty() ||
            emptyStateMessage.fetchSemanticsNode(false) != null
        }
        return this
    }

    fun assertOrdersDisplayed(): SellerOrdersPage {
        assert(orderCards.fetchSemanticsNodes().isNotEmpty()) {
            "Orders should be displayed"
        }
        return this
    }

    fun assertEmptyState(): SellerOrdersPage {
        emptyStateMessage.assertExists()
        return this
    }

    fun assertOrderCount(expectedCount: Int): SellerOrdersPage {
        val actualCount = orderCards.fetchSemanticsNodes().size
        assert(actualCount == expectedCount) {
            "Should have $expectedCount orders, found $actualCount"
        }
        return this
    }

    fun assertOrderDetailDisplayed(): SellerOrdersPage {
        orderIdLabel.assertExists() or orderTotalLabel.assertExists()
        return this
    }

    fun assertAcceptButtonVisible(): SellerOrdersPage {
        acceptButton.assertExists()
        return this
    }

    fun assertMarkPreparingButtonVisible(): SellerOrdersPage {
        markPreparingButton.assertExists()
        return this
    }

    fun assertMarkReadyButtonVisible(): SellerOrdersPage {
        markReadyButton.assertExists()
        return this
    }
}
